### PR TITLE
Support authentication tokens for uploading to Hackage

### PR DESCRIPTION
**Pull request by SebTee**
Adds support for authenticating uploads using authentication tokens, as discussed in issue #6738.

A new flag `--token` (`-t`) has been created. Token authentication takes precedence over username and password meaning that, if a token is set, the username and password flags are ignored.

The token is included in a HTTP header as follows: `Authorization: X-ApiKey [token]`.

## QA Notes

These tests must be run using `http-transport:` `curl`, `wget`, `powershell`, and `plain-http`.

- Calling `cabal upload -t [valid-token] [./path/to/package.tar.gz]` should successfully upload the package to Hackage.
  - The user shouldn't be prompted to enter a username and password.

- Calling `cabal upload -t [invalid-token] [./path/to/package.tar.gz]` should fail to upload the package to Hackage.
  - A HTTP 401 error should be returned with the message "Bad auth token".
  - The user shouldn't be prompted to enter a username and password.

- Calling `cabal upload -t [invalid-token] -u [username] -p [password] [./path/to/package.tar.gz]` should fail to upload the package to Hackage.
  - Authentication token takes precedence over username and password.
  - A HTTP 401 error should be returned with the message "Bad auth token".

- Calling `cabal upload -t` should output a cabal error message reading ``Error: cabal: option `-t' requires an argument TOKEN``.

- Calling `cabal upload [./path/to/package.tar.gz]` should default to username and password authentication.
  - If username and password are not set in the config file, prompt the user for a username and password.
  - If username and password are set in the config file, cabal will attempt to upload the package using the username and password for authentication. 

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by SebTee** [comment]
Token authentication is working for `cabal upload`. I'm currently working on getting token authentication working for `cabal report`.

File: doc/cabal-commands.rst [hunk 1]
```diff
@@ -1076,6 +1076,10 @@ to Hackage.
     documentation for a published package (and not a candidate), add
     ``--publish``.
 
+.. option:: -t, --token
+
+    Your Hackage authentication token.
```
Comment (ffaf1) [inline doc/cabal-commands.rst:1081]: Is there any page we can link? (Ideally something like “How to generate and use auth tokens on hackage”)

Comment (SebTee) [reply doc/cabal-commands.rst:1081]: You can create and delete tokens on your own manage account page. The following link will prompt you to log in to Hackage where it tells you how to create and delete tokens: https://hackage.haskell.org/users/account-management. I've looked through Hackage but I haven't found any mention of token authentication outside of the logged in account management page.  Would it be okay linking to a page that requires a login in this situation?

Comment (ffaf1) [reply doc/cabal-commands.rst:1081]: I would say so, something like

```
Your Hackage authentication token. See XYZ on how to create/delete tokens.
```

Thanks!

**Comment by SebTee** [comment]
I've been having a couple issues with `cabal report`. It doesn't read the token, username or password from the config, you have to set them in the command line. This is also the case for username and password in the cabal-install version 3.10.1.0 release. I've not been able to find an issue for this, does one exist?

The other issue I'm encountering is that I can't seem to work out how to create a build report that `cabal report` can parse and upload to Hackage. My changes so far allow you to set a token that will be passed to the `postHttp` function if a build report is found and parsed. But, since I don't have a parsable build report to upload, I've not been able to fully test this.

**Comment by andreabedini** [comment]
IIRC there's a hidden and undocumented `remote-build-reporting` field. You can find it in the code: https://github.com/search?q=repo%3Ahaskell%2Fcabal%20remote-build-reporting&type=code

**Comment by gbaz** [comment]
I think its fine to not worry about testing the `report` feature. It was sort of stillborn if I recall. I double checked and the hackage builders generates reports as part of the build, but uses its own custom upload machinery.

File: cabal-install/src/Distribution/Client/Config.hs [hunk 2]
```diff
@@ -568,6 +572,7 @@ instance Semigroup SavedConfig where
         UploadFlags
           { uploadCandidate = combine uploadCandidate
           , uploadDoc = combine uploadDoc
+          , uploadToken = combine uploadToken
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/Config.hs:575]: what happens if both fields are non empty?

File: cabal-install/src/Distribution/Client/HttpUtils.hs [hunk 3]
```diff
@@ -535,12 +536,13 @@ curlTransport prog =
...
-            (Just (uname, passwd), _) -> Just (uname ++ ":" ++ passwd)
-            (Nothing, Just a) -> Just a
-            (Nothing, Nothing) -> Nothing
-      case mbAuthString of
-        Just up ->
+      let mbAuthStringToken = case (explicitAuth, uriDerivedAuth) of
+            (Just (Right token), _) -> Just $ Right token
+            (Just (Left (uname, passwd)), _) -> Just $ Left (uname ++ ":" ++ passwd)
+            (_, Just a) -> Just $ Left a
+            (_, _) -> Nothing
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/HttpUtils.hs:543]: Not happy with catch-alls here, they could (in theory) cause problems in the future. Is there a reason why we should not put `Nothing` there?

File: cabal-install/src/Distribution/Client/HttpUtils.hs [hunk 4]
```diff
@@ -1073,7 +1080,9 @@ plainHttpTransport =
           setOutHandler (debug verbosity)
           setUserAgent userAgent
           setAllowBasicAuth False
-          setAuthorityGen (\_ _ -> return auth)
+          case auth of
+            Just (Left x) -> setAuthorityGen (\_ _ -> return $ Just x)
+            _ -> setAuthorityGen (\_ _ -> return Nothing)
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/HttpUtils.hs:1085]: Again this picks up `Just (Right...)` and `Nothing`. Are we sure this is correct (and will be correct for the foreseeable future?)

File: cabal-install/src/Distribution/Client/Setup.hs [hunk 5]
```diff
@@ -1675,6 +1677,17 @@ reportCommand =
     , commandDefaultFlags = defaultReportFlags
     , commandOptions = \_ ->
         [ optionVerbosity reportVerbosity (\v flags -> flags{reportVerbosity = v})
+        , option
+            ['t']
+            ["token"]
+            "Hackage Token."
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/Setup.hs:1683]: Maybe “Hackage authentication token”?

File: cabal-install/src/Distribution/Client/Types/Credentials.hs [hunk 6]
```diff
@@ -1,9 +1,15 @@
...
+  ( Auth
+  , Token (..)
+  , Username (..)
   , Password (..)
   ) where
 
 import Prelude (String)
+import Data.Either (Either)
 
+type Auth = Either (String, String) String
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/Types/Credentials.hs:11]: Maybe add a line of comment to explain what this is.

File: cabal-install/src/Distribution/Client/Types/Credentials.hs [hunk 7]
```diff
@@ -1,9 +1,15 @@
 module Distribution.Client.Types.Credentials
-  ( Username (..)
+  ( Auth
+  , Token (..)
+  , Username (..)
   , Password (..)
   ) where
 
 import Prelude (String)
+import Data.Either (Either)
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/Types/Credentials.hs:9]: Is this not automatically imported from cabal prelude? And if not, why not import it in the line above?

File: cabal-install/src/Distribution/Client/HttpUtils.hs (hunk 4 above)
Comment (SebTee) [reply cabal-install/src/Distribution/Client/HttpUtils.hs:1093]: There was an issue with the plain-http transport that I have now fixed (along with the wget and powershell transports) but this code is left unchanged. Plain-http has to have username and password set using a function passed to `setAuthorityGen` in the browser. The token has to be set as a header in the request. If the token is set then there won't be a security challenge meaning the function passed to `setAuthorityGen` will never be called by the browser.

File: cabal-install/src/Distribution/Client/Config.hs (hunk 2 above)
Comment (SebTee) [reply cabal-install/src/Distribution/Client/Config.hs:575]: When multiple tokens are set in the config file it uses the last token. If a token is set in the command line it takes precedence over the token in the config file. If multiple tokens are set in the command line, like in the config file, the last one is used. If both fields are non empty it chooses the last field and discards the first.

**Review by geekosaur — APPROVED** [review]

**Comment by Mikolaj** [comment]
@gbaz: could (one of the) Hackage maintainers review this PR as well? BTW, I understand that Hackage is immutable, so the QA tests need to upload real sensible packages, not test packages, do they? Or is it fine, because no `--publish` is used and so these would be candidates? But if so, does it test enough? Are things likely to break more with `--publish`?

**Comment by gbaz** [comment]
Its fine to test this without --publish, and just upload candidates.

**Comment by Mikolaj** [comment]
OK, fine.

I'm re-running the failed CI jobs, since the failure is probably due to the unix package breakage.

**Comment by Mikolaj** [comment]
CI passed. @SebTee: once you are satisfied with the state of the PR, please kindly add label squash+merge_me and in 2 days of inactivity the PR is going to be merged.

**Pull request merged**